### ISSUE TITLE
avoid chaining hooks; fix error handling for existing cases

### ISSUE THREAD

**Issue by floatplane** [stevekuznetsov/avalanche-forecast#120]
Fixed a couple of these in #117, but https://github.com/stevekuznetsov/avalanche-forecast/blob/dfce2c50130e4f292ba2cfa9b3fb8f2e876c8f72/hooks/useMapViewZones.ts still has this problem - not sure if there are others, as well.

general rule:
- if a hook ends up with a call to `useQuery`, then that should be the only hook referenced
- the query fn passed to `useQuery` has a lot of latitude in terms of calling async functions - just write out an async function that calls other async methods in sequence, and use that

**Comment by stevekuznetsov** [comment]
Is there a good blog or set of blogs on composing hooks? Seems weird that you wouldn't be able to do that cleanly.

**Comment by floatplane** [comment]
Haven't seen any blogs, this is just stuff I figured out through staring at code.

I'd say the problem is really `useQuery`, as opposed to hooks generally. Or maybe I'd say the problem is directly exposing the return type of `useQuery` as part of the hook. Like if you have a setup like this you're going to be unhappy (and we had this pattern in a couple of places):

```js
export const useTwoQueries = () => {
  const {
    data: hello,
  } = useQuery<string, Error>(['key_one'], () => {
    if (Math.random() > 0.5) {
      throw new Error('oh no');
    } else {
      return 'hello';
    }
  });

  return useQuery<number, Error>(['key_two'], async () => {
    // sleep for 5 seconds
    await new Promise(resolve => setTimeout(resolve, 5000));
    return hello.length;
  },
  {enabled: !!hello});
};
```

1. The way this hook is written, if the first query fails, the second one just hangs in the `idle` state forever
2. You can naively think that you should return the result of the first query if it fails, but then the return type of useTwoQueries changes based on what happens. Also, early returns + hooks will cause something to whine at you 😄 
3. The return type of `useTwoQueries` as it's written is `UseQueryResult<number, Error>`, but that's not covariant with the result of the first call to `useQuery`, and getting them aligned requires a lot of manual property copying.

An approach that probably also works is to make the return value of the hook something explicit, as opposed to just returning the result of the final `useQuery`: 

```js
export const useTwoQueries = () => {
  const {
    isLoading: isLoadingOne,
    isError: isErrorOne,
    data: dataOne,
  } = useQuery<string, Error>(['key_one'], () => {
    if (Math.random() > 0.5) {
      throw new Error('oh no');
    } else {
      return 'hello';
    }
  });

  const {
    isLoading: isLoadingTwo,
    isError: isErrorTwo,
    data: dataTwo,
  } = useQuery<number, Error>(['key_two'], async () => {
    // sleep for 5 seconds
    await new Promise(resolve => setTimeout(resolve, 5000));
    return dataOne.length;
  },
  {enabled: !!dataOne});

  return {
    isLoading: isLoadingOne || isLoadingTwo,
    isError: isErrorOne || isErrorTwo,
    data: dataTwo,
  }
};
```

**Comment by stevekuznetsov** [comment]
@floatplane this looks like the authoritative suggestion  (at least as of the time of writing...) from the author of the library: https://github.com/TanStack/query/discussions/889#discussioncomment-105882

Yikes.

**Comment by stevekuznetsov** [comment]
@floatplane one issue with the bundled `await`s is that you don't end up caching those intermediate requests, and you don't get deduping, etc. I'm thinking let's hoist all the data fetches into hooks and up to the top level in the components and I'll write a nicer wrapper to handle all the intermediate loading and error states.

**Comment by stevekuznetsov** [comment]
Kind of an enormous bummer that this is not easier. I feel like I'm saying that more and more lately ...

**Comment by stevekuznetsov** [comment]
Fixed in #275
